tlv.go

```diff
@@ -0,0 +1,153 @@
+// Type-Length-Value splitting and parsing for proxy protocol V2
+// See spec https://www.haproxy.org/download/1.8/doc/proxy-protocol.txt sections 2.2 to 2.7 and
+// Amazon's extension https://docs.aws.amazon.com/elasticloadbalancing/latest/network/load-balancer-target-groups.html#proxy-protocol
+
+package proxyproto
+
+import (
+	"encoding/binary"
+	"errors"
+	"io"
+	"io/ioutil"
+)
+
+const (
+	// Section 2.2
+	PP2_TYPE_ALPN           PP2Type = 0x01
+	PP2_TYPE_AUTHORITY              = 0x02
+	PP2_TYPE_CRC32C                 = 0x03
+	PP2_TYPE_NOOP                   = 0x04
+	PP2_TYPE_SSL                    = 0x20
+	PP2_SUBTYPE_SSL_VERSION         = 0x21
+	PP2_SUBTYPE_SSL_CN              = 0x22
+	PP2_SUBTYPE_SSL_CIPHER          = 0x23
+	PP2_SUBTYPE_SSL_SIG_ALG         = 0x24
+	PP2_SUBTYPE_SSL_KEY_ALG         = 0x25
+	PP2_TYPE_NETNS                  = 0x30
+	// Section 2.2.7, reserved types
+	PP2_TYPE_MIN_CUSTOM     = 0xE0
+	PP2_TYPE_MAX_CUSTOM     = 0xEF
+	PP2_TYPE_MIN_EXPERIMENT = 0xF0
+	PP2_TYPE_MAX_EXPERIMENT = 0xF7
+	PP2_TYPE_MIN_FUTURE     = 0xF8
+	PP2_TYPE_MAX_FUTURE     = 0xFF
+	// Amazon's extension
+	PP2_TYPE_AWS            = 0xEA
+	PP2_SUBTYPE_AWS_VPCE_ID = 0x01
+)
+
+var (
+	ErrTruncatedTLV    = errors.New("Truncated TLV")
+	ErrMalformedTLV    = errors.New("Malformed TLV Value")
+	ErrIncompatibleTLV = errors.New("Incompatible TLV type")
+)
+
+// PP2Type is the proxy protocol v2 type
+type PP2Type byte
+
+// TLV is a uninterpreted Type-Length-Value for V2 protocol, see section 2.2
+type TLV struct {
+	Type   PP2Type
+	Length int
+	Value  []byte
+}
+
+// readTLVs reads the Type-Length-Value vector, returns the vector or an error.  Reads until EOF or an error occurs.
+func readTLVs(reader io.Reader) ([]TLV, error) {
+	rest, err := ioutil.ReadAll(reader)
+	if err != nil {
+		return nil, err
+	}
+	var tlvs []TLV
+	for i := 0; i < len(rest); {
+		tlv := TLV{
+			Type: PP2Type(rest[i]),
+		}
+		tlv.Length = int(binary.BigEndian.Uint16(rest[i+1 : i+3])) // Max length = 65K
```

Comment:
What happens if `i+1` or `i+3` is outside of index range of `rest`? Would probably be worthwhile to have some unit tests around edge cases.

Reply:
Great catch, added check, will add test.

Reply:
Added.  I actually couldn't get this to fail because the underlying capacity was so large (from `ioutil.ReadAll`) that it wasn't OOBing.